Refactor gulp build process, address file name versioning development issue

sfount opened the pull request:
This pull request updates `gulpfile.js` to accomplish four main goals: 
### 1. Developement `watch` fix

**Fix issue that causes version file names to grow too large and break the `gulp watch` process** 
:tada: :+1: :1st_place_medal: :fireworks: :champagne: 

### 2. Simplify process 

```js
gulp.task('client', ['js', 'css', 'less', 'i18n', 'vendor', 'static'], templateHTML);
```

Make the process easier to understand, the current version in master includes many different styles and quirks of gulp. This refactor follows the linting process and attempts to do everything in the same `gulp.src` `pipe` style.

There is nothing special about the client build, it just processes javascript, css, less, translation files, vendor files and static assets. It then links all versioned build files by templating them into HTML. 

### 3. Improve versioning system 

We use `gulp-rev` to generate hashes of files and include these in filenames in order to 'cache-bust' in development and production. This has been updated and standardised so that every part of the build process can version itself and update the collective manifest file. 

It also updates the cleaning system so that versioned files are removed in favour of the most up to date copy.

### 4. Standardise process 

Move towards standardising the build process to allow future upgrades. [(upcoming gulp 4 release)](https://github.com/gulpjs/gulp/tree/4.0)

### Dependency changes: 
* `rimraf` - library deprecated in favour of `del` [(github)](https://github.com/robrich/gulp-rimraf)
* `gulp-rev-replace` - library deprecated in favour of `gulp-rev-rewrite` [(github)](https://github.com/jamesknelson/gulp-rev-replace)
* `merge-stream` - library added [(github)](https://github.com/grncdr/merge-stream)


This change updates dependencies so a `yarn install` is required. 

This PR should be thoroughly tested by the development team and preferably in production scenarios before being merged with production code.

(The offending `watch` problem:)
![screenshot from 2018-09-02 19-29-42](https://user-images.githubusercontent.com/2844572/44959474-a3e1de80-aee6-11e8-95cb-eaa0eba1fb0e.png)

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/1328

sfount committed:
fix(build): removed infinite concat versioning

This commit introduces brand new gulp build methods that use latest
dependencies and incrementally ensure that the file concat issue that
would cause gulp watch to break previously have been removed.

sfount committed:
refactor(build): clean up gulp file, update deps.

This commit updates dev dependencies, removing unused gulp libraries and
adding newer replacements.

It also patches client unit tests to use the manfiest file and get the
latest built files and introduces top level aliases for common gulp
tasks.

jniles reviewed (approved):
Looks good to me! We've needed a good re-write of the gulpfile for a while.  It seems to have about the same performance on my local machine.

bors r+

sfount committed:
fix(build): watch static template files

This commit ensures that static template files (only those included in
the bhima repository, not vendor files) are tracked through the `watch`
task.

mbayopanda commented on client/src/index.html:
```diff
@@ -10,10 +10,10 @@
   <link rel="stylesheet" href="vendor/angular-ui-grid/ui-grid.min.css">
   <link rel="stylesheet" href="vendor/components-font-awesome/css/font-awesome.min.css">
-  <link rel="stylesheet" href="css/style.min.css">
+  <link rel="stylesheet" href="css/bhima.min.css">
```
It seems like we have some bhima css files inside `/css` folder and other outside the `/css` folder, it will be good to have all bhima css files inside one folder such as `/css` or `/css/bhima/` or flattened into one file `bhima.min.css`.

sfount replied:
This should be addressed with the same logic as https://github.com/IMA-WorldHealth/bhima-2.X/pull/3123#discussion_r215196781.

mbayopanda commented on client/src/index.html:
```diff
@@ -104,8 +104,8 @@ <h4><a class="title-content" ui-sref="index">{{ AppCtrl.project.name }}</a></h4>
-<script src="js/vendor.min.js"></script>
-<script src="js/bhima.min.js"></script>
+<script src="vendor.min.js"></script>
+<script src="js/dist/bhima.min.js"></script>
```
`bhima.min.js` is the endpoint of all angular code of bhima, so we can put it in the same level (the root folder) as `vendor.min.js`.

All we need in `/bin/client` as javascript code are `vendor.min.js` and `bhima.min.js`, so there is no need to have different folders.

sfount replied:
@mbayopanda thanks for the comment! I will update these to be clearer for the build process. 

The organisation of these files isn't as important as `client/src` but should still be considered.

The reason for the different folder is because of the 'cleanup' process. The last build process simply deleted the entire `bin/client` folder on the `client-build` and didn't delete anything on the `watch` method. This iteration tries to delete smaller folders only when they are used, for example `clean-js` will only delete the `bin/client/js` folder when a new `bhima.min.js` is being generated.  For this reason nothing else was but in the `bin/client/js` folder because it would have been deleted at the time. 

I propose to try and keep the targetted folder clean style but factor in your comments we make these files build into seperate js subfolder, for example: 

```
- bin/ 
- - bin/client 
- - - bin/client/js
- - - - bin/client/js/bhima/
- - - - - bhima.min.js
- - - - bin/client/js/vendor/
- - - - - vendor.min.js
```

The same goes for CSS as per your previous comment.

mbayopanda commented on gulpfile.js:
```diff
@@ -58,86 +38,61 @@ const paths = {
       '!client/src/i18n/**/*.js',
     ],
-    css         : ['client/src/css/*.css'],
-    vendorStyle : [
```
👍

sfount committed:
chore(build): standardise js and css build folders

This commit updates the `js` and `css` tasks to standardise the path
names for a clearer `index.html`

mbayopanda reviewed (approved):
LGTM 👍 

bors  r+
